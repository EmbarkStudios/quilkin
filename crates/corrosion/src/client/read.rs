//! Deserialization of changes sent from a corrosion agent

pub use corro_api_types::{QueryEvent, SqliteValue};
use eyre::ContextCompat as _;
use quilkin_types::{AddressKind, Endpoint, IcaoCode, TokenSet};
use serde::{
    Deserialize,
    de::{self, SeqAccess},
};
use std::{collections::BTreeSet, fmt, str::FromStr};

pub trait FromSqlValue: Sized {
    fn from_sql(values: &[SqliteValue]) -> eyre::Result<Self>;
}

#[derive(Debug, PartialEq)]
pub struct ServerRow {
    pub endpoint: Endpoint,
    pub icao: IcaoCode,
    pub tokens: TokenSet,
}

pub fn deserialize_token_set(s: &str) -> eyre::Result<TokenSet> {
    let mut ts = BTreeSet::default();

    let mut tokens = data_encoding::BASE64_NOPAD.decode(s.as_bytes())?;

    if tokens.is_empty() {
        return Ok(TokenSet(ts));
    }

    if tokens[0] & 0x80u8 != 0 {
        let len = (tokens[0] & !0x80) as usize;
        for tok in tokens[1..].chunks_exact(len) {
            ts.insert(tok.to_vec());
        }
    } else if tokens[0] > 1 {
        let mut toks = &tokens[1..];
        for _ in 0..tokens[0] as usize {
            let len = toks[0] as usize;
            eyre::ensure!(
                len <= toks.len() - 1,
                "token length {len} is longer than remaining binary slice"
            );

            ts.insert(toks[1..1 + len].to_vec());
            toks = &toks[1 + len..];
        }
    } else {
        tokens.remove(0);
        ts.insert(tokens);
    }

    Ok(TokenSet(ts))
}

#[inline]
pub fn parse_endpoint(addr: &str) -> eyre::Result<Endpoint> {
    let (addr, port) = addr.rsplit_once(':').context("missing ':'")?;
    let port = port.parse()?;
    if let Some(ip) = addr.strip_prefix('|') {
        let ip = ip.parse()?;
        Ok(Endpoint::new(AddressKind::Ip(ip), port))
    } else {
        Ok(Endpoint::new(AddressKind::Name(addr.to_owned()), port))
    }
}

macro_rules! get_column {
    ($index:expr, $name:literal, $v:expr) => {
        $v.get($index)
            .context(concat!("missing column '", $name, "'"))?
            .as_str()
            .context(concat!("column '", $name, "' is not a string"))?
    };
}

macro_rules! get_json {
    ($name:literal, $conv:expr, $seq:expr) => {{
        let v = $seq
            .next_element()?
            .ok_or(de::Error::missing_field($name))?;
        $conv(v).map_err(de::Error::custom)?
    }};
}

impl FromSqlValue for ServerRow {
    fn from_sql(values: &[SqliteValue]) -> eyre::Result<Self> {
        let endpoint = parse_endpoint(get_column!(0, "endpoint", values))?;
        let icao = get_column!(1, "icao", values).parse()?;
        let tokens = deserialize_token_set(get_column!(2, "tokens", values))?;

        Ok(Self {
            endpoint,
            icao,
            tokens,
        })
    }
}

impl<'de> Deserialize<'de> for ServerRow {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct Visitor;

        impl<'de> de::Visitor<'de> for Visitor {
            type Value = ServerRow;

            fn expecting(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                f.write_str("array of server columns")
            }

            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
            where
                A: SeqAccess<'de>,
            {
                let endpoint = get_json!("endpoint", parse_endpoint, seq);
                let icao = get_json!("icao", IcaoCode::from_str, seq);
                let tokens = get_json!("tokens", deserialize_token_set, seq);

                // Ignore the rest of the elements, if we don't we'll leave
                // the deserializer with tokens that will cause an error
                while let Some(Ignore) = seq.next_element()? {}

                Ok(ServerRow {
                    endpoint,
                    icao,
                    tokens,
                })
            }
        }

        deserializer.deserialize_seq(Visitor)
    }
}

pub struct Ignore;

impl<'de> de::Visitor<'de> for Ignore {
    type Value = Self;

    fn expecting(&self, _formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        Ok(())
    }

    #[inline]
    fn visit_bool<E>(self, _x: bool) -> Result<Self::Value, E> {
        Ok(Self)
    }

    #[inline]
    fn visit_i64<E>(self, _x: i64) -> Result<Self::Value, E> {
        Ok(Self)
    }

    #[inline]
    fn visit_i128<E>(self, _x: i128) -> Result<Self::Value, E> {
        Ok(Self)
    }

    #[inline]
    fn visit_u64<E>(self, _x: u64) -> Result<Self::Value, E> {
        Ok(Self)
    }

    #[inline]
    fn visit_u128<E>(self, _x: u128) -> Result<Self::Value, E> {
        Ok(Self)
    }

    #[inline]
    fn visit_f64<E>(self, _x: f64) -> Result<Self::Value, E> {
        Ok(Self)
    }

    #[inline]
    fn visit_str<E>(self, _s: &str) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        Ok(Self)
    }

    #[inline]
    fn visit_none<E>(self) -> Result<Self::Value, E> {
        Ok(Self)
    }

    #[inline]
    fn visit_some<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: de::Deserializer<'de>,
    {
        Self::deserialize(deserializer)
    }

    #[inline]
    fn visit_newtype_struct<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: de::Deserializer<'de>,
    {
        Self::deserialize(deserializer)
    }

    #[inline]
    fn visit_unit<E>(self) -> Result<Self::Value, E> {
        Ok(Self)
    }

    #[inline]
    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
    where
        A: de::SeqAccess<'de>,
    {
        while let Some(Self) = seq.next_element()? {}
        Ok(Self)
    }

    #[inline]
    fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
    where
        A: de::MapAccess<'de>,
    {
        while let Some((Self, Self)) = map.next_entry()? {}
        Ok(Self)
    }

    #[inline]
    fn visit_bytes<E>(self, _bytes: &[u8]) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        Ok(Self)
    }

    fn visit_enum<A>(self, data: A) -> Result<Self::Value, A::Error>
    where
        A: serde::de::EnumAccess<'de>,
    {
        use serde::de::VariantAccess;
        data.variant::<Self>()?.1.newtype_variant()
    }
}

impl<'de> Deserialize<'de> for Ignore {
    #[inline]
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: de::Deserializer<'de>,
    {
        deserializer.deserialize_ignored_any(Self)
    }
}
