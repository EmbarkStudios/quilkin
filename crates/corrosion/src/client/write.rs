//! Serialization of queries and transactions sent to a corrosion agent

use crate::{
    Peer,
    api::{SqliteParam, Statement},
};
use quilkin_types::{AddressKind, Endpoint, IcaoCode, TokenSet};

pub trait ToSqlParam {
    fn to_sql(&self) -> SqliteParam;
}

pub type Statements<const N: usize> = smallvec::SmallVec<[Statement; N]>;

impl ToSqlParam for TokenSet {
    /// Converts a token set to a SQL parameter
    ///
    /// Due to the limitations imposed on us via JSON (binary data is cumbersome) and SQLite (no arrays)
    /// we base64 a custom encoding for token sets
    fn to_sql(&self) -> SqliteParam {
        const MAX_TOKENS: usize = u8::MAX as usize >> 1;
        let tokens = &self.0;
        if tokens.is_empty() {
            return SqliteParam::Null;
        }

        let mut blob = smallvec::SmallVec::<[u8; 512]>::new();

        // We could varint encode this instead, but for now just fail
        debug_assert!(
            tokens.len() <= MAX_TOKENS,
            "number of tokens ({}) is more than {MAX_TOKENS}",
            tokens.len()
        );

        let len_prefix = if tokens.len() > 1 {
            // If all the tokens have the same length, and that length is less than
            // MAX_TOKENS, we can skip length prefixing each token
            let len = tokens.first().unwrap().len();
            let same_len = tokens.iter().all(|tok| tok.len() == len);

            if same_len && len <= MAX_TOKENS {
                blob.push(0x80 | len as u8);
            } else {
                blob.push(tokens.len() as u8);
            }

            !same_len
        } else {
            blob.push(1);
            false
        };

        for tok in tokens {
            if len_prefix {
                debug_assert!(
                    tok.len() <= u8::MAX as usize,
                    "token length {} is more than {}",
                    tok.len(),
                    u8::MAX
                );

                blob.push(tok.len() as u8);
            }

            blob.extend_from_slice(&tok);
        }

        SqliteParam::Text(data_encoding::BASE64_NOPAD.encode(&blob).into())
    }
}

impl ToSqlParam for IcaoCode {
    fn to_sql(&self) -> SqliteParam {
        SqliteParam::Text(self.as_str().into())
    }
}

impl ToSqlParam for Endpoint {
    fn to_sql(&self) -> SqliteParam {
        SqliteParam::Text(to_compact_str(self))
    }
}

impl ToSqlParam for std::net::Ipv6Addr {
    fn to_sql(&self) -> SqliteParam {
        SqliteParam::Text(compact_str::format_compact!("{self}"))
    }
}

#[inline]
fn to_compact_str(ep: &Endpoint) -> compact_str::CompactString {
    use std::fmt::Write as _;

    let mut cs = compact_str::CompactString::default();

    match &ep.address {
        AddressKind::Name(hn) => {
            cs.push_str(hn);
        }
        AddressKind::Ip(ip) => {
            // Put a | which is invalid in both hostnames and IPs so that
            // when parsing we can easily distinguish IPs from hostnames
            write!(&mut cs, "|{ip}").unwrap();
        }
    }

    write!(&mut cs, ":{}", ep.port).unwrap();
    cs
}

pub struct Server<'s, const N: usize> {
    pub peer: Peer,
    pub statements: &'s mut smallvec::SmallVec<[Statement; N]>,
}

impl<'s, const N: usize> Server<'s, N> {
    #[inline]
    pub fn for_peer(peer: Peer, statements: &'s mut smallvec::SmallVec<[Statement; N]>) -> Self {
        Self { peer, statements }
    }

    /// Create a statement to insert a new server
    #[inline]
    pub fn upsert(&mut self, endpoint: &Endpoint, icao: IcaoCode, tokens: &TokenSet) {
        let mut params = Vec::with_capacity(4);

        params.push(endpoint.to_sql());
        params.push(icao.to_sql());
        params.push(tokens.to_sql());

        let peer_ip = self.peer.ip().to_string();

        self.statements.push(Statement::WithParams(
            format!("INSERT INTO servers (endpoint,icao,tokens,contributors,cont_update) VALUES (?,?,?,jsonb('{{\"{peer_ip}\":{{}}}}'),unixepoch('now'))
             ON CONFLICT(endpoint) DO UPDATE SET
                contributors = jsonb_patch(contributors,'{{\"{peer_ip}\":{{}}}}'),
                cont_update = unixepoch('now')
             WHERE excluded.icao = servers.icao"),
            params,
        ));

        let server = endpoint.address.to_string();

        self.statements.push(Statement::WithParams(
            format!(
                "INSERT INTO dc (ip,port,icao,servers) VALUES (?,?,?,jsonb('{{\"{server}\":{{}}}}'))
            ON CONFLICT(ip) DO UPDATE SET
                servers = jsonb_patch(servers,'{{\"{server}\":{{}}}}')
            WHERE excluded.icao = dc.icao"
            ),
            vec![peer_ip.into(), self.peer.port().into(), icao.to_sql()],
        ));
    }

    /// Create a statement to remove the specified server immediately
    ///
    /// Unlike [`Self::remove_deferred`], deletion will occur regardless of how
    /// many contributors there are to the server
    #[inline]
    pub fn remove_immediate(&mut self, endpoint: &Endpoint) {
        self.statements.push(Statement::WithParams(
            "DELETE FROM servers WHERE rowid = (SELECT MIN(rowid) FROM servers WHERE endpoint = ?)"
                .into(),
            vec![endpoint.to_sql()],
        ));

        let server = endpoint.address.to_string();

        self.statements.push(Statement::WithParams(
            format!("UPDATE dc SET servers = jsonb_patch(servers, '{{\"{server}\":null}}') WHERE rowid = (SELECT MIN(rowid) FROM dc WHERE ip = ?)"),
            vec![self.peer.ip().to_string().into()]
        ));
    }

    /// Create a statement to remove the peer as a contributor to the server
    ///
    /// This method won't immediately delete the server like [`Self::remove_immediately`]
    /// but will update it to remove the specified peer as a contributor. The server
    /// can be removed later if it no longer has contributors after a specified
    /// time period.
    #[inline]
    pub fn remove_deferred(&mut self, endpoint: &Endpoint) {
        let peer_ip = self.peer.ip().to_string();

        self.statements.push(Statement::WithParams(
            format!(
                "UPDATE servers SET
                contributors = jsonb_patch(contributors,'{{\"{peer_ip}\":null}}'),
                cont_update = unixepoch('now')
            WHERE rowid = (SELECT MIN(rowid) FROM servers WHERE endpoint = ?)"
            ),
            vec![endpoint.to_sql()],
        ));

        let server = to_compact_str(endpoint);

        self.statements.push(Statement::WithParams(
            format!("UPDATE dc SET servers = jsonb_patch(servers, '{{\"{server}\":null}}') WHERE rowid = (SELECT MIN(rowid) FROM dc WHERE ip = ?)"),
            vec![peer_ip.into()]
        ));
    }

    /// Create a statement to update one or more server columns
    pub fn update(&mut self, update: UpdateBuilder<'_>) {
        let mut query = String::with_capacity(128);
        query.push_str("UPDATE servers SET ");

        let mut params = Vec::with_capacity(update.params() + 1);

        if let Some(icao) = update.icao {
            query.push_str("icao = ?");
            params.push(SqliteParam::Text(icao.as_str().into()));
        }

        if let Some(ts) = update.tokens {
            if !params.is_empty() {
                query.push_str(", ");
            }

            query.push_str("tokens = ?");
            params.push(ts.to_sql());
        }

        // We know we are only updating one row, so ideally we would just stick
        // LIMIT 1 at the end...unfortunately we can't. SQLite only supports LIMIT
        // on UPDATE queries when built with `SQLITE_ENABLE_UPDATE_DELETE_LIMIT`
        // ...but that doesn't work https://github.com/rusqlite/rusqlite/issues/1111
        query.push_str(" WHERE rowid = (SELECT MIN(rowid) FROM servers WHERE endpoint = ?)");
        params.push(update.ep.to_sql());

        self.statements.push(Statement::WithParams(query, params));
    }

    /// Create a statement to remove servers with no contributors whose last
    /// update was older
    ///
    /// Note that unlike the other methods, the peer for this does not matter
    #[inline]
    pub fn reap_old(&mut self, max_age: std::time::Duration) {
        self.statements.push(Statement::Simple(format!(
            "DELETE FROM servers WHERE length(contributors) <= 1 AND unixepoch('now') - cont_update > {}", max_age.as_secs()
        )));
    }
}

pub struct UpdateBuilder<'s> {
    ep: &'s Endpoint,
    icao: Option<IcaoCode>,
    tokens: Option<&'s TokenSet>,
}

impl<'s> UpdateBuilder<'s> {
    #[inline]
    pub fn new(ep: &'s Endpoint) -> Self {
        Self {
            ep,
            icao: None,
            tokens: None,
        }
    }

    #[inline]
    pub fn update_icao(mut self, icao: IcaoCode) -> Self {
        self.icao = Some(icao);
        self
    }

    #[inline]
    pub fn update_tokens(mut self, ts: &'s TokenSet) -> Self {
        self.tokens = Some(ts);
        self
    }

    #[inline]
    fn params(&self) -> usize {
        let mut count = 0;
        if self.icao.is_some() {
            count += 1
        }
        if self.tokens.is_some() {
            count += 1
        }
        count
    }
}

impl ToSqlParam for Peer {
    fn to_sql(&self) -> SqliteParam {
        use std::fmt::Write as _;
        let mut cs = compact_str::CompactString::default();
        write!(&mut cs, "{}", self.ip()).unwrap();
        SqliteParam::Text(cs)
    }
}

pub struct Datacenter<'s, const N: usize>(pub &'s mut smallvec::SmallVec<[Statement; N]>);

impl<'s, const N: usize> Datacenter<'s, N> {
    #[inline]
    pub fn insert(&mut self, peer: Peer, qcmp: u16, icao: IcaoCode) {
        let mut params = Vec::with_capacity(3);

        params.push(peer.to_sql());
        params.push(SqliteParam::Integer(qcmp as _));
        params.push(icao.to_sql());

        self.0.push(Statement::WithParams(
            "INSERT INTO dc (ip,port,icao,servers) VALUES (?,?,?,jsonb('{}'))".into(),
            params,
        ));
    }

    /// Create a statement to remove the specified peer
    ///
    /// The peer is also removed as a contributor for all servers it still knows
    /// of, to be cleaned up later if the server has no contributors
    ///
    /// The time for the update can be specified, defaulting to `UtcDateTime::now`
    /// if not specified
    #[inline]
    pub fn remove(&mut self, peer: Peer, update_time: Option<time::UtcDateTime>) {
        let time = update_time.unwrap_or(time::UtcDateTime::now());

        self.0.push(Statement::WithParams(format!(
            "WITH sj AS (SELECT server.key FROM dc JOIN json_each(dc.servers) AS server WHERE ip = '{0}' LIMIT 1)
            UPDATE servers SET
                contributors = jsonb_patch(s.contributors,'{{\"{0}\":null}}'),
                cont_update = {1}
            FROM servers s
            LEFT JOIN sj ON s.endpoint = sj.key", peer.ip(), time.unix_timestamp()
        ), vec![]));

        self.0.push(Statement::WithParams(
            "DELETE FROM dc WHERE rowid = (SELECT MIN(rowid) FROM dc WHERE ip = ?)".into(),
            vec![peer.to_sql()],
        ));
    }

    /// Create a statement to update one or more datacenter columns
    pub fn update(&mut self, peer: Peer, port: Option<u16>, icao: Option<IcaoCode>) {
        debug_assert!(port.is_some() || icao.is_some());

        let mut query = String::with_capacity(128);
        query.push_str("UPDATE dc SET ");

        let mut params = Vec::with_capacity(3);

        if let Some(port) = port {
            query.push_str("port = ?");
            params.push(SqliteParam::Integer(port as _));
        }

        if let Some(icao) = icao {
            if !params.is_empty() {
                query.push_str(", ");
            }
            query.push_str("icao = ?");
            params.push(icao.to_sql());
        }

        query.push_str(" WHERE rowid = (SELECT MIN(rowid) FROM dc WHERE ip = ?");
        params.push(peer.to_sql());

        self.0.push(Statement::WithParams(query, params));
    }
}

pub struct Filter<'s, const N: usize>(pub &'s mut smallvec::SmallVec<[Statement; N]>);

impl<'s, const N: usize> Filter<'s, N> {
    #[inline]
    pub fn upsert(&mut self, filter: &str) {
        self.0.push(Statement::WithParams(
            "INSERT INTO filter (id,filter) VALUES (9999,?) ON CONFLICT(id) DO UPDATE SET filter = excluded.filter".into(),
            vec![SqliteParam::Text(filter.into())]
        ));
    }
}
