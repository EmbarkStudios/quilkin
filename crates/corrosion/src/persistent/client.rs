use bytes::Bytes;
use corro_api_types::ExecResponse;
use quilkin_types::IcaoCode;
use std::net::SocketAddr;
use tokio::sync::{mpsc, oneshot};

type ResponseTx = oneshot::Sender<Result<ExecResponse, StreamError>>;

#[derive(thiserror::Error, Debug)]
pub enum StreamError {
    #[error(transparent)]
    Connect(#[from] quinn::ConnectionError),
    #[error(transparent)]
    Write(#[from] quinn::WriteError),
    #[error(transparent)]
    Read(#[from] quinn::ReadError),
    #[error(transparent)]
    ReadExact(#[from] quinn::ReadExactError),
    #[error(transparent)]
    Reset(#[from] quinn::ResetError),
    #[error(transparent)]
    Json(serde_json::Error),
    #[error(
        "expected a chunk of JSON length {} but received {}",
        expected,
        received
    )]
    LengthMismatch { expected: usize, received: usize },
    #[error("stream ended")]
    StreamEnded,
}

use super::LengthReadError as Lre;

impl From<Lre> for StreamError {
    fn from(value: Lre) -> Self {
        match value {
            Lre::Json(json) => Self::Json(json),
            Lre::LengthMismatch { expected, received } => {
                Self::LengthMismatch { expected, received }
            }
            Lre::ReadExact(re) => Self::ReadExact(re),
            Lre::Read(r) => Self::Read(r),
            Lre::StreamEnded => Self::StreamEnded,
        }
    }
}

#[derive(thiserror::Error, Debug)]
pub enum ConnectError {
    #[error(transparent)]
    Connect(#[from] quinn::ConnectError),
    #[error(transparent)]
    Connection(#[from] quinn::ConnectionError),
    #[error(transparent)]
    Creation(#[from] std::io::Error),
    #[error(transparent)]
    Handshake(#[from] super::HandshakeError),
    #[error(transparent)]
    Write(#[from] StreamError),
}

#[derive(thiserror::Error, Debug)]
pub enum TransactionError {
    #[error(transparent)]
    Json(#[from] serde_json::Error),
    #[error(transparent)]
    Stream(#[from] StreamError),
    #[error("the I/O task for this client was shutdown")]
    TaskShutdown,
}

/// The current version of the client stream
///
/// - 0: Invalid
/// - 1: The initial version
///   Requests are 16-bit length-prefixed JSON, where the JSON is [`ServerChange`]
///   Responses are the JSON of [`ExecResult`]
pub const VERSION: u16 = 1;

/// A persistent connection to a corrosion agent
pub struct Client {
    inner: quinn::Connection,
    local_addr: SocketAddr,
    tx: mpsc::UnboundedSender<(Bytes, ResponseTx)>,
    task: tokio::task::JoinHandle<Result<Option<quinn::VarInt>, StreamError>>,
}

impl Client {
    /// Connects using a non-encrypted session
    pub async fn connect_insecure(
        addr: SocketAddr,
        qcmp_port: u16,
        icao: IcaoCode,
    ) -> Result<Self, ConnectError> {
        let ep = quinn::Endpoint::client((std::net::Ipv6Addr::LOCALHOST, 0).into())?;

        let inner = ep
            .connect_with(
                quinn_plaintext::client_config(),
                addr,
                &addr.ip().to_string(),
            )?
            .await?;

        // This is really infallible
        let local_addr = ep.local_addr()?;

        let client = inner.clone();
        let (mut send, mut recv) = client.open_bi().await?;

        // Handshake
        // We need to actually send something for the connection to be fully established
        let peer_version = {
            let req = super::ClientHandshakeRequestV1 { qcmp_port, icao }.write();

            send.write_chunk(super::write_length_prefixed(&req).freeze())
                .await
                .map_err(StreamError::from)?;

            let res = super::read_length_prefixed(&mut recv)
                .await
                .map_err(StreamError::from)?;
            match super::ServerHandshake::read(VERSION, &res[..])? {
                super::ServerHandshake::V1(shs) => {
                    if !shs.accept {
                        return Err(ConnectError::Handshake(
                            crate::persistent::HandshakeError::UnsupportedVersion {
                                ours: VERSION,
                                theirs: 1,
                            },
                        ));
                    }

                    1
                }
            }
        };

        let (tx, mut reqrx) = mpsc::unbounded_channel();

        let task = tokio::task::spawn(async move {
            let func = async || -> Result<Option<quinn::VarInt>, StreamError> {
                match peer_version {
                    1 => loop {
                        let (msg, comp): (_, ResponseTx) = tokio::select! {
                            res = recv.received_reset() => {
                                return res.map_err(StreamError::Reset);
                            }
                            req = reqrx.recv() => {
                                let Some(req) = req else {
                                    let _ = send.reset(quinn::VarInt::from_u32(1));
                                    let _ = send.finish();
                                    // We need to drop the recv stream so that the server
                                    // knows we don't care and it can finish closing the connection
                                    drop(recv);
                                    tracing::debug!("waiting for server to received buffered stream...");
                                    drop(send.stopped().await);
                                    tracing::debug!("client finished");
                                    break;
                                };

                                req
                            }
                        };

                        send.write_chunk(msg).await?;
                        let res = super::read_length_prefixed_jsonb::<ExecResponse>(&mut recv)
                            .await
                            .map_err(StreamError::from);

                        if let Err(error) = &res {
                            tracing::error!(%error, "error occurred reading response to transaction");
                        }

                        if comp.send(res).is_err() {
                            tracing::warn!("transaction response could not be sent to queuer");
                        }
                    },
                    _invalid => {
                        return Err(StreamError::Connect(
                            quinn::ConnectionError::VersionMismatch,
                        ));
                    }
                }

                Ok(None)
            };

            func().await
        });

        Ok(Self {
            inner,
            tx,
            task,
            local_addr,
        })
    }

    pub fn local_addr(&self) -> SocketAddr {
        self.local_addr
    }

    pub fn remote_addr(&self) -> SocketAddr {
        self.inner.remote_address()
    }

    pub async fn transactions(
        &self,
        change: &[super::ServerChange],
    ) -> Result<ExecResponse, TransactionError> {
        let buf = super::write_length_prefixed_jsonb(&change)?;

        let (tx, rx) = oneshot::channel();
        self.tx
            .send((buf.freeze(), tx))
            .map_err(|_| TransactionError::TaskShutdown)?;
        Ok(rx.await.map_err(|_| TransactionError::TaskShutdown)??)
    }

    /// Closes the connection to the upstream server
    pub async fn shutdown(self) {
        drop(self.tx);
        if let Ok(Err(error)) = self.task.await {
            tracing::warn!(%error, "stream exited with error");
        }
        drop(self.inner);
    }
}
