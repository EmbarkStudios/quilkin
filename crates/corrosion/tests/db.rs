//! Tests the DB mutations and queries work as expected

use corro_api_types::SqliteValue;
use corro_types::{agent::SplitPool, api::Statement};
use corrosion::client::read::{FromSqlValue, ServerRow};
use corrosion_utils as tu;
use quilkin_types::{AddressKind, Endpoint, IcaoCode};
use std::net::{Ipv4Addr, Ipv6Addr, SocketAddrV6};

async fn exec_all<const N: usize>(v: &mut smallvec::SmallVec<[Statement; N]>, sp: &SplitPool) {
    let mut conn = sp.write_priority().await.unwrap();
    let tx = conn.transaction().unwrap();
    tu::exec(&tx, v.iter()).unwrap();
    tx.commit().unwrap();
    v.clear();
}

async fn read_server_row(id: usize, sp: &SplitPool) -> ServerRow {
    let conn = sp.read().await.unwrap();
    conn.query_row(
        "SELECT endpoint,icao,tokens FROM servers WHERE rowid = ?",
        [id],
        |row| {
            let mut v = Vec::with_capacity(3);
            v.push(row.get::<_, SqliteValue>(0).unwrap());
            v.push(row.get::<_, SqliteValue>(1).unwrap());
            v.push(row.get::<_, SqliteValue>(2).unwrap());
            Ok(ServerRow::from_sql(&v).unwrap())
        },
    )
    .unwrap()
}

fn make_row(i: u32) -> ServerRow {
    let address = match i % 3 {
        0 => AddressKind::Ip(Ipv4Addr::from_bits(i).into()),
        1 => AddressKind::Name(format!("boop.{i}.net")),
        2 => AddressKind::Ip(Ipv6Addr::from_bits(i as _).into()),
        _ => unreachable!(),
    };

    let endpoint = Endpoint {
        address,
        port: i as u16,
    };

    ServerRow {
        endpoint,
        icao: IcaoCode::new_testing([b'B', b'O', b'O', b'P']),
        tokens: [i.to_ne_bytes()].into(),
    }
}

const PREP_PEER: SocketAddrV6 = SocketAddrV6::new(Ipv6Addr::from_bits(0xaaffeeff), 8999, 0, 0);

async fn prep(name: &str) -> SplitPool {
    let sp = tu::new_split_pool(name, corrosion::schema::SCHEMA).await;

    const MAX: usize = 100;

    let mut iv = smallvec::SmallVec::<[_; MAX]>::new();
    let mut s = corrosion::client::write::Server::for_peer(PREP_PEER, &mut iv);

    for i in 0..1000u32 {
        let row = make_row(i);

        if s.statements.len() == MAX {
            exec_all(s.statements, &sp).await;
        }

        s.upsert(&row.endpoint, row.icao, &row.tokens);
    }

    if !s.statements.is_empty() {
        exec_all(s.statements, &sp).await;
    }

    sp
}

/// Tests basics of inserting servers and datacenters
#[tokio::test]
async fn inserts_and_reads_servers() {
    let sp = prep("inserts_and_reads_servers").await;

    {
        let r = sp.read().await.unwrap();
        assert_eq!(
            1000,
            r.query_row("SELECT COUNT(*) FROM servers", [], |r| r.get::<_, u32>(0))
                .unwrap()
        );
    }

    for i in 0..3u32 {
        let row = read_server_row(i as usize + 1, &sp).await;
        let expected = make_row(i);

        assert_eq!(row, expected);
    }
}

/// Tests that servers that have no datacenter contributors are reaped after
/// some amount of time
#[tokio::test]
async fn collects_old_servers() {
    let sp = prep("collects_old_servers").await;
    let fake_time = time::UtcDateTime::now() - std::time::Duration::from_secs(60 * 60);

    {
        let r = sp.read().await.unwrap();
        assert_eq!(
            1000,
            r.query_row("SELECT COUNT(*) FROM servers", [], |r| r.get::<_, u32>(0))
                .unwrap()
        );
    }

    let mut v = smallvec::SmallVec::<[_; 2]>::new();

    // Remove the DC as a contributor, but set the time of the update to an hour in the past
    {
        let mut dc = corrosion::client::write::Datacenter(&mut v);
        dc.remove(PREP_PEER, Some(fake_time));

        let mut conn = sp.write_priority().await.unwrap();
        let tx = conn.transaction().unwrap();
        tu::exec(&tx, dc.0.iter()).unwrap();
        tx.commit().unwrap();
        dc.0.clear();
    }

    {
        let r = sp.read().await.unwrap();
        assert_eq!(
            1000,
            r.query_row("SELECT COUNT(*) FROM servers", [], |r| r.get::<_, u32>(0))
                .unwrap()
        );
    }

    // Add a new server that should not be deleted since it still has a contributor
    {
        let mut s = corrosion::client::write::Server::for_peer(PREP_PEER, &mut v);
        s.upsert(
            &Endpoint {
                address: AddressKind::Ip(Ipv6Addr::from_bits(0x888888888888).into()),
                port: 8888,
            },
            IcaoCode::new_testing([b'V'; 4]),
            &[8888u64.to_ne_bytes()].into(),
        );

        let mut conn = sp.write_priority().await.unwrap();
        let tx = conn.transaction().unwrap();
        tu::exec(&tx, s.statements.iter()).unwrap();
        tx.commit().unwrap();
        s.statements.clear();
    }

    // Do the actual removal of the servers with no contributors that are older than 30 minutes
    {
        let mut s = corrosion::client::write::Server::for_peer(PREP_PEER, &mut v);
        s.reap_old(std::time::Duration::from_secs(60 * 30));

        let mut conn = sp.write_priority().await.unwrap();
        let tx = conn.transaction().unwrap();
        tu::exec(&tx, s.statements.iter()).unwrap();
        tx.commit().unwrap();
    }

    let only_row = {
        let conn = sp.read().await.unwrap();
        let statement = conn
            .prepare("SELECT endpoint,icao,tokens,json(contributors) FROM servers")
            .unwrap();
        tu::query_to_string(statement, |sql, row| {
            row.add_cell(tu::Cell::new(
                &corrosion::client::read::parse_endpoint(&sql.get::<_, String>(0).unwrap())
                    .unwrap()
                    .to_string(),
            ));
            row.add_cell(tu::Cell::new(&sql.get::<_, String>(1).unwrap()));
            row.add_cell(tu::Cell::new(&format!(
                "{:?}",
                corrosion::client::read::deserialize_token_set(&sql.get::<_, String>(2).unwrap())
                    .unwrap()
            )));
            row.add_cell(tu::Cell::new(
                &serde_json::from_str::<serde_json::Value>(&sql.get::<_, String>(3).unwrap())
                    .unwrap()
                    .to_string(),
            ));
        })
    };

    insta::assert_snapshot!("only_one", only_row);
}
