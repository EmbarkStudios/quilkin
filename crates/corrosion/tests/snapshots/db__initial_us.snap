---
source: crates/corrosion/tests/db.rs
expression: only_row().await
---
+-----------+------+--------------------------+
| endpoint  | icao | tokens                   |
+===========+======+==========================+
| 0.0.0.0:0 | BOOP | TokenSet({[0, 0, 0, 0]}) |
+-----------+------+--------------------------+
